module/zfs/sa.c

```diff
@@ -451,6 +451,54 @@ sa_add_layout_entry(objset_t *os, sa_attr_type_t *attrs, int attr_count,
 	return (tb);
 }
 
+static sa_lot_t *
+sa_lookup_cached_layout(sa_os_t *sa, uint64_t hash, sa_attr_type_t *attrs,
+    int count)
+{
+	sa_layout_cache_t *slc;
+	sa_lot_t *tb = NULL;
+	int i, j;
+
+	kpreempt_disable();
+	slc = sa->sa_layout_cache + CPU_SEQID;
+
+	for (i = 0; i < SA_LAYOUT_ENTRIES; i++) {
+		if (slc->slc_lot[i] == NULL)
+			break;
+		if (slc->slc_lot[i]->lot_hash != hash)
+			continue;
+		if (sa_layout_equal(slc->slc_lot[i], attrs, count) == 0) {
+			tb = slc->slc_lot[i];
+			break;
+		}
+	}
+	if (tb != NULL) {
+		/* shift toward the most recently used */
+		j = (i + 1) % SA_LAYOUT_ENTRIES;
+		if (j != slc->slc_head) {
+			sa_lot_t *t = slc->slc_lot[i];
+			ASSERT(j < SA_LAYOUT_ENTRIES);
+			slc->slc_lot[i] = slc->slc_lot[j];
+			slc->slc_lot[j] = t;
+		}
+	}
+	kpreempt_enable();
+
+	return (tb);
+}
+
+static void
+sa_add_cached_layout(sa_os_t *sa, sa_lot_t *tb)
+{
+	sa_layout_cache_t *slc;
+
+	kpreempt_disable();
+	slc = sa->sa_layout_cache + CPU_SEQID;
+	slc->slc_lot[slc->slc_head] = tb;
+	slc->slc_head = (slc->slc_head + 1) % SA_LAYOUT_ENTRIES;
+	kpreempt_enable();
+}
+
 static void
 sa_find_layout(objset_t *os, uint64_t hash, sa_attr_type_t *attrs,
```

Comment:
@bzzz77 it's possible we could get it in to the next rounds of updates if you can provide a patch.